File: reactapp/src/types/substitution/bartering/barteringExercise.ts
import { IMerchant, Merchant } from "./merchant.ts";
import { LinearEquation } from "@/types/math/linearEquation.ts";
import { Trade } from "./trade.ts";
import { Type } from "class-transformer";

export interface IBarteringExercise {
    readonly id: number;
    readonly linearEquation: LinearEquation;
    readonly firstMerchant: IMerchant;
    readonly secondMerchant: IMerchant;
    readonly thirdMerchant: IMerchant;
}

export class BarteringExercise implements IBarteringExercise {
    public readonly id: number;

    @Type(() => LinearEquation)
    public readonly linearEquation: LinearEquation;

    @Type(() => Merchant)
    public readonly firstMerchant: Merchant;

    @Type(() => Merchant)
    public readonly secondMerchant: Merchant;

    @Type(() => Merchant)
    public readonly thirdMerchant: Merchant;

    public trade!: Trade;

    constructor(id: number, linearEquation: LinearEquation, firstMerchant: Merchant, secondMerchant: Merchant, thirdMerchant: Merchant) {
        this.id = id;
        this.linearEquation = linearEquation;
        this.firstMerchant = firstMerchant;
        this.secondMerchant = secondMerchant;
        this.thirdMerchant = thirdMerchant;
    }

    public setTrade(): void {
        this.trade = Trade.computeTradeFromEquation(this.linearEquation);
    }
}

File: webapi/Models/Substitution/Bartering/BarteringExercise.cs
using webapi.Models.ConceptualKnowledge;
using webapi.Models.Math;

namespace webapi.Models.Substitution.Bartering
{
    public class BarteringExercise : ICKExercise
    {
        public long? Id { get; set; }

        public int? Ordering { get; set; }

        public int? Level { get; set; }

        public required LinearEquation LinearEquation { get; set; }

        public required Merchant FirstMerchant { get; set; }

        public required Merchant SecondMerchant { get; set; }

        public required Merchant ThirdMerchant { get; set; }
    }
}
